File: 29Items-Projects/7Algorithms/2GRAPH-ALGORITHMS/java/src/main/java/com/example/graphalgorithms/Bridges.java
package com.example.graphalgorithms;

import java.util.*;

public class Bridges {
    public static class Bridge {
        private final String from;
        private final String to;

        public Bridge(String from, String to) {
            this.from = from;
            this.to = to;
        }

        public String from() { return from; }
        public String to() { return to; }
        public String toString() { return "(" + from + " - " + to + ")"; }
    }
    private int time = 0;

    public static List<Bridge> find(Map<String, List<String>> graph) {
        Bridges solver = new Bridges();
        Set<String> visited = new HashSet<>();
        Map<String, Integer> discovery = new HashMap<>();
        Map<String, Integer> low = new HashMap<>();
        Map<String, String> parent = new HashMap<>();
        List<Bridge> bridges = new ArrayList<>();
        for (String vertex : graph.keySet()) {
            if (!visited.contains(vertex)) {
                solver.dfs(graph, vertex, visited, discovery, low, parent, bridges);
            }
        }
        return bridges;
    }

    private void dfs(
        Map<String, List<String>> graph,
        String vertex,
        Set<String> visited,
        Map<String, Integer> discovery,
        Map<String, Integer> low,
        Map<String, String> parent,
        List<Bridge> bridges
    ) {
        visited.add(vertex);
        discovery.put(vertex, time);
        low.put(vertex, time);
        time++;
        for (String neighbor : graph.getOrDefault(vertex, List.of())) {
            if (!visited.contains(neighbor)) {
                parent.put(neighbor, vertex);
                dfs(graph, neighbor, visited, discovery, low, parent, bridges);
                low.put(vertex, Math.min(low.get(vertex), low.get(neighbor)));
                if (low.get(neighbor) > discovery.get(vertex)) {
                    bridges.add(new Bridge(vertex, neighbor));
                }
            } else if (!neighbor.equals(parent.get(vertex))) {
                low.put(vertex, Math.min(low.get(vertex), discovery.get(neighbor)));
            }
        }
    }

    public static void main(String[] args) {
        System.out.println("Bridges");
        System.out.println(find(Map.of("A", List.of("B"), "B", List.of("A", "C", "D"), "C", List.of("B"), "D", List.of("B"))));
    }
}

File: 29Items-Projects/7Algorithms/2GRAPH-ALGORITHMS/python/graph_algorithms/bridges.py
from __future__ import annotations


def bridges(graph: dict[str, list[str]]) -> list[tuple[str, str]]:
    visited: set[str] = set()
    discovery: dict[str, int] = {}
    low: dict[str, int] = {}
    parent: dict[str, str | None] = {}
    result: list[tuple[str, str]] = []
    time = 0

    def dfs(vertex: str) -> None:
        nonlocal time
        visited.add(vertex)
        discovery[vertex] = low[vertex] = time
        time += 1
        for neighbor in graph.get(vertex, []):
            if neighbor not in visited:
                parent[neighbor] = vertex
                dfs(neighbor)
                low[vertex] = min(low[vertex], low[neighbor])
                if low[neighbor] > discovery[vertex]:
                    result.append((vertex, neighbor))
            elif neighbor != parent.get(vertex):
                low[vertex] = min(low[vertex], discovery[neighbor])

    for vertex in graph:
        if vertex not in visited:
            parent[vertex] = None
            dfs(vertex)
    return result


if __name__ == "__main__":
    sample = {"A": ["B"], "B": ["A", "C", "D"], "C": ["B"], "D": ["B"]}
    print("Bridges")
    print(bridges(sample))

